Connection Status: Added Network Broadcast Receiver

nitish1211 opened the pull request:
Fixes #312

I've added the the snackbar which shows up whenever there is change in network state.
The snackbar showing offline status stays there till connection is re-established.
Once the connection is established the snackbar showing connection established dismisses after a short interval.

@kunall17 @Sam1301  Can you please review this. And I am working on #82 , so will update the PR soon.

smarx commented:
*Automated message from Dropbox CLA bot*

@nitish1211, it looks like you've already signed the Dropbox CLA. Thanks!

kunall17 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2141,4 +2143,32 @@ public void onClick(View view) {
+                boolean isConnected = intent.getBooleanExtra(ConnectivityManager.EXTRA_NO_CONNECTIVITY, false);
+                if(isConnected){
+                    //Disable fab when there is no connectivity
+                    fab.setEnabled(false);
```
What if the chatBox is displayed currently, there should be a check if that's visible close that one as well!

nitish1211 replied:
Here I was thinking that the chat box would change to fab on scrolling and once it changes to fab we cant go back to chat box as fab is disabled.

kunall17 replied:
But there can be case where the user is being displayed the chatbox and a network change occurs, 
Maybe just disable the asyncSend button also save the state, like if the user network disconnects and they were writing a message so we can show the chatbox again when the network connects, for a better UX

nitish1211 replied:
This is how it currently is working @kunall17  Please suggest any changes that need to be added.

![g_20170123_0106563](https://cloud.githubusercontent.com/assets/13676490/22249952/19107d6e-e26b-11e6-96bf-c0c9834130d8.gif)

The displaying of the snackbar at the bottim of the screen takes the chatbox out of the view and as soon as the network connection is reestablished the text that the user had entered is still present.

kunall17 replied:
Hmmm, something is going wrong mostly the snackbar moves the FAB above to make space for itself and not appear on top of it, can you check why?

nitish1211 replied:
Its because we have a set a custom `app:layout_behaviour` in **main.xml** for our fab which does not include moving up on snackbar's appearance. The moving up on snackbar's appearance is included in the default behaviour of a fab though this can be fixed by creating a different custom layout behaviour class.

kunall17 commented:
How does the snackbar looks in the night mode?

nitish1211 commented:
Looks fine 

![screenshot_20170121-093359](https://cloud.githubusercontent.com/assets/13676490/22171865/0bef524a-dfbe-11e6-829d-10a88eec64af.png)

kunall17 commented:
Also unregister the receiver like https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/activities/ZulipActivity.java#L1998

nitish1211 commented:
The BroadcastReceiver and `IntentFilter filter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);`  are working fine I've tested it on a API 24 device and on API 23 and 19 on the emulator.

ShailyTyagi commented:
Even when the device is offline the snackbar displays 'connection esatblished' .I ran it on an API level 22 device.
![2017_01_21_23_01_55](https://cloud.githubusercontent.com/assets/18248854/22176360/77908a3a-e02e-11e6-875e-8bc82aa9278d.gif)

ShailyTyagi commented:
![2017_01_21_23_01_55](https://cloud.githubusercontent.com/assets/18248854/22176403/8634dfd6-e02f-11e6-980e-881b8333c188.gif)
even when the device is offline the snackbar shows 'connection established'.

kunall17 commented:
Also you have only checked if we are connected to wifi, It would be better for checking if the internet is working while connected to wifi!

nitish1211 commented:
Here I've added a new `layout_behaviour` file for fab. The fab now animates in and out of view like this


![g_20170124_2135569](https://cloud.githubusercontent.com/assets/13676490/22255431/5cbedc24-e27d-11e6-8262-8eb57fd6148e.gif)

Also added a line to hide the chat box when network is lost.
There is also a call to `isnetworkAvailable()` to check if there is network available.
The BroadcastReceiver has also been unregistered during onPause() and registered onResume().

nitish1211 commented:
I've also disabled the refresh button in the overflow menu when the device goes offline.
![screenshot_20170127-010301](https://cloud.githubusercontent.com/assets/13676490/22347248/4a10ff74-e42d-11e6-8bbb-984035508afe.png)


And I'm working on the offline display of messages #82 . I'll send a separate PR for that.
@kunall17 @Sam1301  Can you please review the changes done till now, would really like any suggestions from your side.

kunall17 commented:
@nitish1211 Its better to include #82 in this PR only as they both are connected and will be easier to debug! 
For this #82 you just need to play with fetch() method here https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/activities/MessageListFragment.java#L244 this is how we fetch messages!

nitish1211 commented:
I've added Offline compatibility now the messages stored offline load into the view.
Here's how it looks like now:

![g_20170129_1206563](https://cloud.githubusercontent.com/assets/13676490/22402407/d3d110b0-e61b-11e6-9630-8b5e34a67082.gif)


I have used the fetch method by giving a call to in ready to display.

kunall17 commented:
@nitish1211 awesome! 
split the commit which deals with offline messages display!

nitish1211 commented:
@kunall17 Done.

kunall17 commented:
The android docs say about the isConnected() 
> Indicates whether network connectivity exists and it is possible to establish connections and pass data.

I'm still not clear by the docs in a scenario where the wifi is connected by the internet isn't working, what happens in this case?

kunall17 commented:
I tried this branch I kept my wifi on but did not start the zulip dev server! 
And it showed me connection established! 

Same goes in the case where my wifi is connected it will show the same messages! 

![device-2017-01-31-234441](https://cloud.githubusercontent.com/assets/12700799/22478324/6a3c4930-e80f-11e6-825c-693aff9301c5.png)
Still overlaps the FAB

Just an suggestion: try to test with the error thrown 
> java.net.ConnectException: Failed to connect to /10.0.3.2:9991

nitish1211 commented:
@kunall17 You are right, the `isConnected()` only checks for an active network but does not check if there is access to the internet. I think the solution would be to send an `httpget` request to google.com(or any url just to test the connection). It would raise an Exception if there is no internet connectivity. 

For the Fab that was happening because I had tried to combine both the default behaviour(Going out of view when snackbar appears) of the fab and hiding it on scroll into one class which was causing the error. It can be fixed by changing the behaviour of the fab in the `ZulipActivity.java` whenever there is change in network state.

yadav-rahul commented:
@nitish1211 Can you resolve merge conflicts?

nitish1211 commented:
@yadav-rahul  Resolved the merge conflicts. Sorry it took so long as I was travelling.

nitish1211 commented:
@kunall17 Ive added the http request that checks whether the internet connection is active and during that duration displays **Connecting ....** snackbar and changes to Connected or No Connection depending on whether the http request went through.

@kunall17 @Sam1301 @yadav-rahul @vishwesh3  Please review this and please let me know if you can suggest any changes to make it better.

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -1929,6 +1938,20 @@ public boolean onCreateOptionsMenu(Menu menu) {
+    @Override
+    public boolean onPrepareOptionsMenu(Menu menu) {
+        if(isNetworkAvailable())
+        {
```
Code formatting.

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
         RESET_DATABASE,
     }
+
+    public void setupNetworkBroadcastReceiver()
```
Add JAVADOC what this function does.

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
     }
+
+    public void setupNetworkBroadcastReceiver()
+    {
```
Code formatting

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
+                    onReadyToDisplay(true);
+                    Snackbar.make(coordinatorLayout,R.string.no_connection,Snackbar.LENGTH_INDEFINITE).show();
+                    Log.d("Network Listener", "No Internet");
+                }
```
Code formatting

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
+        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
+    }
+
+    private void checkIfOnline(){
```
Add JAVADOC what this function does.

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
+        IntentFilter filter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
+        registerReceiver(networkStateReceiver, filter);
+    }
+    private boolean isNetworkAvailable() {
```
Add JAVADOC what this function does.

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
+                    connection.setConnectTimeout(3000);
+                    connection.connect();
+                } catch (Exception e) {
+                    Log.d("nitish","couldnt connect");
```
Create a TAG filed in the class and use that as a parameter of Log.d

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
+                    Log.d("nitish","couldnt connect");
+                    networkStatus= false;
+                }
+                Log.d("nitish","connected");
```
Create a TAG filed in the class and use that as a parameter of Log.d

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
+            @Override
+            protected void onPostExecute(String s) {
+                Log.d("In Post","execute");
+                if(networkStatus)
```
Code formatting

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2287,4 +2312,88 @@ public void onClick(View view) {
+                    startRequests();
+                    Snackbar.make(coordinatorLayout,R.string.connection_established,Snackbar.LENGTH_SHORT).show();
+                }
+                else
```
Code formatting

vishwesh3 commented on app/src/main/java/com/zulip/android/util/FabRemoveOnSnackbar.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.List;
+
+/**
+ * Created by nitish on 5/2/17.
```
Author details are saved in the git commit author info, you can remove this one!
 Add a comment that what actually this class is doing.

vishwesh3 commented on app/src/main/java/com/zulip/android/util/FabRemoveOnSnackbar.java:
```diff
@@ -0,0 +1,47 @@
+        return false;
+    }
+
+    private float getFabTranslationYForSnackbar(CoordinatorLayout parent,
```
Add JAVADOC what this function does.

nitish1211 commented:
@vishwesh3 Done.

nitish1211 commented:
@kunall17 @vishwesh3 Resolved the merge conflicts. Please review this.

kunall17 commented:
Rather than connecting to google.co.in we can use the request's sent to fetch updates (heartbeat) of the app in AsyncGetEvents! 
And according to there exceptions, timeouts we can display messages and use broadcast when the status has changed, and again fetch a heartbeat and compute the results! 

This way we don't have to request google.com as well!

nitish1211 commented:
@kunall17 Here I've used the zulip's continuous polling to check for network connectivity.
It shows offline when the number of failures to connect is equal to 2(which is equal to a timeout of approx **3 seconds**).

Connectivity is reestablished when there is a certain data response(when `body.getEvents().size() > 0`).

Please suggest any changes required.

kunall17 commented:
- You should not remove the old Behaviour which was being used, edit the current behaviour i found some good article which might help you achieving this https://www.bignerdranch.com/blog/customizing-coordinatorlayouts-behavior/
- Format the code

kunall17 commented on app/src/main/java/com/zulip/android/networking/AsyncGetEvents.java:
```diff
@@ -52,6 +52,9 @@
 public class AsyncGetEvents extends Thread {
     private static final String TAG = "AsyncGetEvents";
     private static final String ASYNC_GET_EVENTS = "asyncGetEvents";
+    private static final String STATUS_CONNECTED = "connected";
```
Declare these constants in the util/Constants.java

nitish1211 commented:
@kunall17 I've added the changes requested by you. Please review them and let me know of any changes you would like to propose.

saketkumar95 commented:
@nitish1211 I tested it on Android 7 and got crash. Swipe the chatBox to dismiss either in online or offline mode.

```
java.lang.ClassCastException: com.zulip.android.util.RemoveFabOnScroll cannot be cast to com.zulip.android.util.RemoveViewsOnScroll
                        at com.zulip.android.activities.ZulipActivity.removeChatBox(ZulipActivity.java:268)
                        at com.zulip.android.util.SwipeRemoveLinearLayout.onInterceptTouchEvent(SwipeRemoveLinearLayout.java:46)
                        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2212)
                        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2669)
                        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
                        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2669)
                        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
                        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2669)
                        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
                        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2669)
                        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
                        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2669)
                        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
                        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2669)
                        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
                        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2669)
                        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
                        at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:411)
                        at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1813)
                        at android.app.Activity.dispatchTouchEvent(Activity.java:3127)
                        at android.support.v7.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:63)
                        at android.support.v7.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:63)
                        at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:373)
                        at android.view.View.dispatchPointerEvent(View.java:10200)
                        at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:4518)
                        at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4349)
                        at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3889)
                        at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3942)
                        at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3908)
                        at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:4035)
                        at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:3916)
                        at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:4092)
                        at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3889)
                        at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3942)
                        at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3908)
                        at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:3916)
                        at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3889)
                        at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:6305)
                        at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:6279)
                        at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:6229)
                        at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:6408)
                        at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:185)
                        at android.os.MessageQueue.nativePollOnce(Native Method)
                        at android.os.MessageQueue.next(MessageQueue.java:323)
                        at android.os.Looper.loop(Looper.java:136)
                        at android.app.ActivityThread.main(ActivityThread.java:6209)
                        at java.lang.reflect.Method.invoke(Native Method)
                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)```

nitish1211 commented:
@saketkumar95 Resolved the Error.

vishwesh3 commented on app/src/main/java/com/zulip/android/networking/AsyncGetEvents.java:
```diff
@@ -282,6 +287,10 @@ public void run() {
+                            //Shows connected status on receiving data
+                            if (failures > 0)
+                                mActivity.showConnectivitySnackBar(Constants.STATUS_CONNECTED);
+                            failures = 0;
```
Why this two times?

vishwesh3 commented on app/src/main/java/com/zulip/android/networking/AsyncGetEvents.java:
```diff
@@ -105,6 +106,10 @@ private void backoff(Exception e) {
         Log.e(ASYNC_GET_EVENTS, "Failure " + failures + ", sleeping for "
                 + backoff);
+        //Two failures represent loss in network connectivity
+        if (failures == 2) {
```
I think this 2 should be defined in `Constant.java`

vishwesh3 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -1972,6 +1965,17 @@ public boolean onCreateOptionsMenu(Menu menu) {
+    @Override
+    public boolean onPrepareOptionsMenu(Menu menu) {
+        if (networkStatus.equals(Constants.STATUS_CONNECTED)) {
+            menu.findItem(R.id.refresh).setEnabled(true);
```
Device is connected to server but **Refresh** option is not enabled. 
![ezgif com-video-to-gif 7](https://cloud.githubusercontent.com/assets/18511177/23401731/ce4c8d8e-fdce-11e6-9756-dad300b72045.gif)

nitish1211 commented:
@vishwesh3 I've solved the issues stated above :)

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -1972,6 +1965,18 @@ public boolean onCreateOptionsMenu(Menu menu) {
 
+    @Override
+    public boolean onPrepareOptionsMenu(Menu menu) {
+        Log.d("nitish","status"+networkStatus);
```
Remove this. @vishwesh3 already told you about this. https://github.com/zulip/zulip-android/pull/316#discussion_r99484688

nitish1211 replied:
I was using it when debugging. Force of habit :wink: will remove it sorry.

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -834,9 +831,6 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
             startActivity(photoSendIntent);
-
-            // activity transition animation
-            ActivityTransitionAnim.transition(ZulipActivity.this);
```
why are you removing this?

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -922,9 +916,6 @@ private void dispatchTakePictureIntent() {
                 startActivityForResult(takePictureIntent, REQUEST_TAKE_PHOTO);
-
-                // activity transition animation
-                ActivityTransitionAnim.transition(ZulipActivity.this);
```
Here also.

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -1109,7 +1100,7 @@ public void onClick(View view) {
     private void showSoftKeyboard() {
         InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
-        imm.toggleSoftInput(InputMethodManager.SHOW_IMPLICIT,InputMethodManager.HIDE_IMPLICIT_ONLY);
+        imm.toggleSoftInput(InputMethodManager.SHOW_IMPLICIT, InputMethodManager.HIDE_IMPLICIT_ONLY);
```
Remove this change.

Sam1301 replied:
I think its probably fine to retain code refactor changes. These occur clearly because someone forgot to refactor code and it was merged. I think it'd be reasonable to include such changes in a different commit :).

nitish1211 replied:
This is because of formatting of the code.Should I include it in a different commit??

Sam1301 replied:
yep :+1:

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -1697,7 +1688,7 @@ public void onBackPressed() {
             return;
         }
-       //check for narrowed or not
+        //check for narrowed or not
```
Remove this too.

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -1752,7 +1743,8 @@ private void dismissToast() {
     /**
      * If previous toast is showing cancel it and show new one
-     * @param string message in the toast
+     *
```
Remove this change.

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -1818,7 +1810,7 @@ private void narrowPMWith(final Person person) {
             list.add(app.getYou());
         doNarrow(new NarrowFilterPM(list));
-        onNarrowFillSendBoxPrivate(new Person[]{person},false);
+        onNarrowFillSendBoxPrivate(new Person[]{person}, false);
```
Remove this too.

nitish1211 replied:
Same case of code formatting :)

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2156,10 +2161,6 @@ private void openLogin(String serverUrl) {
         startActivity(i);
-
-        // activity transition animation
-        ActivityTransitionAnim.transition(this);
```
Why this?

saketkumar95 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2323,7 +2326,7 @@ public void onClick(View view) {
         } else {
             if (messages.length == 1) tempMessage = messages[0];
-            String name = (tempMessage.getType() == MessageType.PRIVATE_MESSAGE) ? getString(R.string.notify_private, tempMessage.getSenderFullName()) : getString(R.string.notify_stream, tempMessage.getStream().getName() , tempMessage.getSubject());
+            String name = (tempMessage.getType() == MessageType.PRIVATE_MESSAGE) ? getString(R.string.notify_private, tempMessage.getSenderFullName()) : getString(R.string.notify_stream, tempMessage.getStream().getName(), tempMessage.getSubject());
```
Remove this change too.

nitish1211 commented:
@saketkumar95 @vishwesh3 @kunall17 Any other changes you guys can suggest ? :smile:

vishwesh3 commented:
I can't find https://github.com/zulip/zulip-android/pull/316#issuecomment-274850058 

Snackbar overlaps the FAB

![148837387481825](https://cloud.githubusercontent.com/assets/18511177/23461108/845c4468-feae-11e6-86e9-ccc32441f928.png)

nitish1211 commented:
There is a new Behaviour file been added[`RemoveFabOnScroll.java`](https://github.com/zulip/zulip-android/pull/316/commits/bcff398e7c2aaa13d7eb7b612b9869b076cacf8c#diff-24420d9290e07f59a3cb142d9a6ebb61) for the Fab which makes sure that the FAB scales out every time the snackbar comes into view.

Its working quite fine for me which android version are you using? @vishwesh3

nitish1211 commented:
This is a demo of how it works currently. :slightly_smiling_face: 
![g_20170301_2039103](https://cloud.githubusercontent.com/assets/13676490/23466005/deed5492-febf-11e6-82bd-3eeedeab3ffa.gif)

saketkumar95 commented:
@nitish1211  When narrowed to any stream , i got this even when i was connected to the internet.

![screen](https://cloud.githubusercontent.com/assets/21558765/23528307/63895f3e-ffbf-11e6-898b-c3ea90264b72.jpeg)

Steps to reproduce : 

Connect to the internet.
Narrow to the any stream.
Disconnect the internet, and reconnect again. 
I waited till 1 min, but the it was showing waiting for network, even through the net was working.

nitish1211 commented:
@saketkumar95 How long did you leave it disconnected?
Because the time to send connect polls increases the longer you leave it offline thus making it take longer to reconnect. 
It increases to **33 seconds** on just the **7th** failure.
The solution to this would be to add a **retry button** in the snackbar. Which would allow the user to send the poll request when he pleases, instead of the app sending it after a set time period.
What are your thoughts on this?

Its working fine but takes a lot of time to reconnect:
![g_20170303_1115523](https://cloud.githubusercontent.com/assets/13676490/23539729/28657a0c-0003-11e7-95f3-3fc630b6d2ee.gif)

nitish1211 commented:
Added a Retry Button and removed the Connection Established SnackBar. I've observed in most of the Google apps that Snackbars are only used to show **loss** of connectivity and are dismissed when connectivity is reestablished.
@saketkumar95 what are your thoughts on this?

So this is how it currently works now:
![g_20170303_1538413](https://cloud.githubusercontent.com/assets/13676490/23547251/84472bc0-0028-11e7-9eb6-905565919b07.gif)

saketkumar95 commented:
@nitish1211 I don't think disabling the refresh button makes sense. If someone is running on slow internet connection , then they might want to refresh the app, but sometimes on slow connection maybe the refresh button gets disabled. This happens sometimes on slow connections.

nitish1211 commented:
That's the reason I've added the **RETRY** button to the snackbar so that people with slow network connections can hit retry to refresh their feed.

saketkumar95 commented:
No need for that. Refresh can do that. I think you should remove retry and enable the refresh again.  If network is not available we can have anythink like toast, snackbar , or an alert dialog with network unavailable message. Alert dialog is used in this cases in whatsapp.

nitish1211 commented:
How about I keep both enabled the refresh and the RETRY button?? 
Makes it faster for the user to refresh feed. :)

saketkumar95 commented:
yeah! That sounds good too. For me it's good to be implemented but lets first confirm with @kunall17 :)

nitish1211 commented:
@saketkumar95 Just figured out that the refresh button should not be kept Enabled when offline, because on pressing the refresh button everything is reset. The _Database is cleared_ and the the coonncetionState is reset. This causes the offline DB to be wiped out. Therefore I believe that the **refresh** button should be kept **disabled** when offline.
